core/domain/stats_services_test.py

```diff
@@ -45,532 +50,765 @@ def _get_continuous_computation_class(cls):
-            'count': 2
-        }])
 
-    def test_get_top_state_rule_answers_from_multiple_explorations(self):
-        exp1 = self.save_new_default_exploration(self.EXP_ID1, self.owner_id)
-        exp_services.update_exploration(self.owner_id, self.EXP_ID1, [{
-            'cmd': exp_domain.CMD_RENAME_STATE,
-            'old_state_name': feconf.DEFAULT_INIT_STATE_NAME,
-            'new_state_name': 'Second',
-        }], 'Change state name')
-        state_name0 = self.state_name
-        state_name1 = exp1.init_state_name
-
-        # There are no initial top answers for these explorations.
-        top_answers_list = stats_services.get_top_state_rule_answers_multi(
-            [(self.EXP_ID0, state_name0), (self.EXP_ID1, state_name1)],
-            [self.DEFAULT_RULESPEC_STR])
-        self.assertEquals(len(top_answers_list), 2)
-        self.assertEquals(len(top_answers_list[0]), 0)
-        self.assertEquals(len(top_answers_list[1]), 0)
-
-        # Submit some answers.
-        self._record_answer(
-            'a', exploration_id=self.EXP_ID0, state_name=state_name0)
-        self._record_answer(
-            'a', exploration_id=self.EXP_ID1, state_name=state_name1)
-        self._record_answer(
-            'b', exploration_id=self.EXP_ID1, state_name=state_name1)
-        self._record_answer(
-            'b', exploration_id=self.EXP_ID1, state_name=state_name1)
-        self._record_answer(
-            'b', exploration_id=self.EXP_ID1, state_name=state_name1)
-        self._record_answer(
-            'c', exploration_id=self.EXP_ID1, state_name=state_name1)
-        self._record_answer(
-            'c', exploration_id=self.EXP_ID0, state_name=state_name0)
-
-        top_answers_list = stats_services.get_top_state_rule_answers_multi(
-            [(self.EXP_ID0, state_name0), (self.EXP_ID1, state_name1)],
-            [self.DEFAULT_RULESPEC_STR])
-        self.assertEquals(len(top_answers_list), 2)
-        self.assertEquals(top_answers_list[0], [{
-            'value': 'a',
-            'count': 1
-        }, {
-            'value': 'c',
-            'count': 1
-        }])
-        self.assertEquals(top_answers_list[1], [{
-            'value': 'b',
-            'count': 3
-        }, {
-            'value': 'a',
-            'count': 1
-        }, {
-            'value': 'c',
-            'count': 1
-        }])
+class AnswerEventTests(test_utils.GenericTestBase):
+    """Test recording new answer operations through events."""
 
-    def test_get_top_state_rule_answers_from_multiple_states(self):
-        state_name0 = self.state_name
-        state_name1 = 'Second'
+    SESSION_ID = 'SESSION_ID'
+    TIME_SPENT = 5.0
+    PARAMS = {}
+
+    def _get_submitted_answer_dict_list(self, state_answers):
+        submitted_answer_dict_list = (
+            state_answers.get_submitted_answer_dict_list())
+        for submitted_answer_dict in submitted_answer_dict_list:
+            del submitted_answer_dict['json_size']
```

Comment:
Because it needs to be stored in the answer dict so that it's persisted in the NDB model. It's used as a mechanism for the answer sharding data structure. I'm happy to explain this in much more detail offline.